docsite/rst/developing_program_flow_modules.rst

```diff
@@ -0,0 +1,392 @@
+.. _flow_modules:
+
+=======
+Modules
+=======
+
+This is an in-depth dive into understanding Ansible's program flow to execute
+modules.  It will be of use to people working on the portions of the Core
+Ansible Engine that execute a module, it may be of interest to people writing
+Ansible Modules, and people simply wanting to use Ansible Modules will likely
+not find anything helpful here.
+
+.. _flow_types_of_modules:
+
+Types of Modules
+================
+
+Ansible supports several different types of modules in its code base.  Some of
+these are for backwards compatibility and others to enable flexibility.
+
+.. _flow_action_plugins:
+
+Action Plugins
+--------------
+
+Action Plugins look like modules to end users who are writing :term:`playbooks` but
+they're distinct entities for the purposes of this paper.  Action Plugins
+always execute on the controller. They sometimes are able to do all work there
+(for instance, the debug Action Plugin which prints some text for the user to
+see or the assert Action Plugin which can test whether several values in
+a playbook satisfy certain criteria.)
+
+More often, Action Plugins set up some values and environment on the controller
```

Comment:
Read my sentence a few times nad decided to remove mention of the environment.  it can fall under "some values" so no need to be explicit about it.  Here's the modified line:

More often, Action Plugins set up some values on the controller, then invoke an
actual module on the managed node that does something with these values.